File: experiments/rl/perception_features.lua
-- Current screen-derived fields and observed history, never internal a-state.
local E=assert(loadfile('training/runtime/rl_perception_execution.lua'))()
local F=assert(loadfile('training/runtime/rl_fighter_perception.lua'))()
local P=assert(loadfile('training/runtime/rl_visible_projectiles.lua'))()
local N={interface='sf2_screen_perception_v2',frame_features=1129,observations=4516}
local function add(out,x)
 assert(type(x)=='number' and x==x and math.abs(x)<math.huge,'Nonfinite perception')
 out[#out+1]=math.max(-1,math.min(1,x))
end
local function hp(p)
 local x=p.displayed_hp
 return type(x)=='number' and x==x and x>=-1 and x<=144 and math.max(0,x)/144 or -1
end
function N.features(s)
 local a,b=s.p1,s.p2;local out={};local va,vb=s.fighter_perception.p1,s.fighter_perception.p2
 local function position(p)
  if p.position_known~=true then return nil end
  local x,y=p.visible_x,p.visible_y
  assert(type(x)=='number' and x==x and math.abs(x)<math.huge)
  assert(type(y)=='number' and y==y and math.abs(y)<math.huge)
  return {x,y}
 end
 local function grounded(p)
  if p.grounded=='unknown' then return -1 end
  assert(type(p.grounded)=='boolean');return p.grounded and 1 or 0
 end
 local pa,pb=position(va),position(vb)
 for _,v in ipairs({hp(a),hp(b),s.timer/99,pa and pb and (pb[1]-pa[1])/384 or -1,
  pa and pa[1]/384 or -1,pb and pb[1]/384 or -1,pa and pa[2]/224 or -1,pb and pb[2]/224 or -1,
  grounded(va),grounded(vb)}) do add(out,v) end
 for i=0,11 do add(out,b.char==i and 1 or 0) end
 for _,block in ipairs({E.features(s),F.features(s),P.features(s)}) do
  for _,v in ipairs(block) do add(out,v) end
 end
 assert(#out==N.frame_features);return out
end
return N

File: experiments/rl/perception_features.py
"""Structured screen observation; never expose hidden action-state one-hots."""
import math
import numpy as np
from .perception_execution import features as execution_features
from .fighter_perception import features as fighter_features
from .visible_projectiles import features as projectile_features

INTERFACE = 'sf2_screen_perception_v2'
FRAME_FEATURES = 1129
OBSERVATIONS = FRAME_FEATURES*4


def health(player):
    value = player.get('displayed_hp')
    # -1 is the game's exhausted displayed-bar sentinel at a mature KO.
    return max(0,value)/144 if type(value) in (int,float) and math.isfinite(value) and -1<=value<=144 else -1.


def features(state):
    a,b = state['p1'],state['p2']
    actors=state['fighter_perception']; va,vb=actors['p1'],actors['p2']
    def position(p):
        if p.get('position_known') is not True:return None
        x,y=p['visible_x'],p['visible_y']
        if any(type(v) not in (int,float) or not math.isfinite(v) for v in (x,y)):
            raise ValueError('Invalid rendered actor position')
        return x,y
    def grounded(p):
        value=p['grounded']
        if value=='unknown':return -1.
        if type(value)is not bool:raise ValueError('Invalid visible grounded state')
        return float(value)
    pa,pb=position(va),position(vb)
    out = [health(a),health(b),state['timer']/99,(pb[0]-pa[0])/384 if pa and pb else -1.,
           pa[0]/384 if pa else -1.,pb[0]/384 if pb else -1.,
           pa[1]/224 if pa else -1.,pb[1]/224 if pb else -1.,grounded(va),grounded(vb)]
    out.extend(float(b['char']==i) for i in range(12))
    out.extend(execution_features(state))
    out.extend(fighter_features(state))
    out.extend(projectile_features(state))
    assert len(out)==FRAME_FEATURES
    array = np.asarray(out,dtype=np.float32)
    if not np.isfinite(array).all(): raise ValueError('Nonfinite screen perception')
    return np.clip(array,-1,1)
